crates/ide_completion/src/render.rs

```diff
@@ -313,17 +306,44 @@ impl<'a> Render<'a> {
     }
 }
 
-fn compute_relevance(ctx: &RenderContext, ty: &Type, name: &str) -> CompletionRelevance {
-    let mut res = CompletionRelevance::default();
+fn compute_exact_type_match(ctx: &CompletionContext, completion_ty: &hir::Type) -> bool {
+    if let Some(expected_type) = ctx.expected_type.as_ref() {
+        // We don't ever consider unit type to be an exact type match, since
+        // nearly always this is not meaningful to the user.
+        completion_ty == expected_type && !expected_type.is_unit()
+    } else {
+        false
+    }
+}
 
-    res.exact_type_match = Some(ty) == ctx.completion.expected_type.as_ref();
-    res.exact_name_match = Some(name) == ctx.completion.expected_name.as_deref();
+fn compute_exact_name_match(ctx: &CompletionContext, completion_name: impl Into<String>) -> bool {
+    let completion_name = completion_name.into();
 
-    res
+    Some(&completion_name) == ctx.expected_name.as_ref()
 }
 
-fn relevance_type_match(db: &dyn HirDatabase, ty: &Type, expected_type: &Type) -> bool {
-    ty == expected_type || ty.autoderef(db).any(|deref_ty| &deref_ty == expected_type)
+fn compute_ref_match(ctx: &CompletionContext, completion_ty: &hir::Type) -> Option<Mutability> {
+    let mut ref_match = None;
+    if let Some(expected_type) = &ctx.expected_type {
+        if completion_ty != expected_type {
+            if let Some(expected_type_without_ref) = expected_type.remove_ref() {
+                if completion_ty == &expected_type_without_ref
+                    || completion_ty
+                        .autoderef(ctx.db)
+                        .any(|deref_ty| deref_ty == expected_type_without_ref)
```

Comment:
```suggestion
                if completion_ty
                        .autoderef(ctx.db)
                        .any(|deref_ty| deref_ty == expected_type_without_ref)
```

Just this might work I think?

bors r+